src/main/java/org/broadinstitute/hellbender/engine/ReferenceDataSourceFromReferenceBases.java

```diff
@@ -0,0 +1,128 @@
+package org.broadinstitute.hellbender.engine;
+
+import htsjdk.samtools.SAMSequenceDictionary;
+import htsjdk.samtools.reference.ReferenceSequence;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.fasta.CachingIndexedFastaSequenceFile;
+import org.broadinstitute.hellbender.utils.iterators.ByteArrayIterator;
+import org.broadinstitute.hellbender.utils.reference.ReferenceBases;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Iterator;
+
+/**
+ * Manages traversals and queries over reference data (for now, fasta files only)
+ *
+ * Supports targeted queries over the reference by interval, but does not
+ * yet support complete iteration over the entire reference.
+ */
+public final class ReferenceDataSourceFromReferenceBases implements ReferenceDataSource {
+
+    private ReferenceBases bases;
+    private SAMSequenceDictionary sequenceDictionary;
+
+    /**
+     * Initialize this data source using ReferenceBases and corresponding sequence dictionary.
+     */
+    public ReferenceDataSourceFromReferenceBases(final ReferenceBases bases, final SAMSequenceDictionary referenceSequenceDictionary) {
+        this.bases = Utils.nonNull(bases);
+        this.sequenceDictionary = referenceSequenceDictionary;
+    }
+
+    /**
+     * Start an iteration over the entire reference. Not yet supported!
+     *
+     * See the BaseUtils class for guidance on how to work with bases in this format.
+     *
+     * @return iterator over all bases in this reference
+     */
+    @Override
+    public Iterator<Byte> iterator() {
+        throw new UnsupportedOperationException("Iteration over entire reference not yet implemented");
+    }
+
+    /**
+     * Query a specific interval on this reference, and get back an iterator over the bases spanning that interval.
+     *
+     * See the BaseUtils class for guidance on how to work with bases in this format.
+     *
+     * @param interval query interval
+     * @return iterator over the bases spanning the query interval
+     */
+    @Override
+    public Iterator<Byte> query( final SimpleInterval interval ) {
+
+        int startIndex = (bases.getInterval().getStart() - interval.getStart());
```

Comment:
Is this backwards? If `bases` starts at position 5, and `interval` starts at position 6 (one base into the bases stored in `bases`), then `startIndex` would be -1. But it should be 1 in that case, right?